### PR TITLE
eframe support for wgpu on the web

### LINKED ISSUE #1755: egui-wgpu on web
How could one use the wgpu integration for egui on wasm? I can't find a way to do this. Is this possible?

### PR DESCRIPTION
**Pull request by Wumpf**
Adds a wgpu painter for eframe's web compilation (i.e. targeting wasm).
You can render the demo with it by replacing `glow` with `wgpu` in the feature list in `build_demo_web.sh`. Worked fine for me on manual testing!

Fixes #1755

Remaining things to figure out (likely after this PR):

* target is currently srgb, this causes blending to be done incorrectly. It seems wgpu doesn't support anything else for webgl canvases right now (why?)
* associate a depth buffer with the egui-wgpu renderer
* expose MSAA settings
* expose limits settings etc.
* allow targeting WebGPU

### CODE REVIEW

**Comment by Wumpf** [comment]
filed wgpu issue for srgb issue https://github.com/gfx-rs/wgpu/issues/3058

**Comment by emilk** [comment]
`cargo check -p egui_demo_app --lib --target wasm32-unknown-unknown` fails with:

```
error[E0597]: `*arg0` does not live long enough
  --> crates/egui_demo_app/src/lib.rs:83:1
   |
83 | #[wasm_bindgen]
   | ^^^^^^^^^^^^^^-
   | |             |
   | |             `*arg0` dropped here while still borrowed
   | borrowed value does not live long enough
   | argument requires that `*arg0` is borrowed for `'static`
   |
   = note: this error originates in the attribute macro `wasm_bindgen` (in Nightly builds, run with -Z macro-backtrace for more info)
```

some shenanigans with how `async` interacts with the `#[wasm_bindgen]` macro

**Comment by emilk** [comment]
`crates/egui_demo_app/src/apps/custom3d_glow.rs` should be made to work before we merge this

**Comment by Wumpf** [comment]
Fixed!
image

File: crates/egui_demo_app/src/apps/custom3d_wgpu_shader.wgsl [hunk 1]
```diff
@@ -8,6 +8,10 @@ struct Uniforms {
     _padding: vec3<f32>, // needed in order to make it 16 byte aligned
 };
 
+struct Uniforms {
+    @size(16) angle: f32, // pad to 16 bytes
+};
```
Comment (emilk) [inline crates/egui_demo_app/src/apps/custom3d_wgpu_shader.wgsl:13]: this looks unused? O.o

**Review by emilk — CHANGES_REQUESTED** [review]

File: crates/egui_demo_app/src/apps/custom3d_wgpu_shader.wgsl (hunk 1 above)
[the remark this answers is not in the document]
Comment (Wumpf) [reply crates/egui_demo_app/src/apps/custom3d_wgpu_shader.wgsl]: oops! Fixed.
wgpu panics for errors and logs everying via logging crate; enabled pushing that to console locally and looked at standalone naga compilation via naga (via https://github.com/PolyMeilex/vscode-wgsl): It seems this isn't a warning right now and I also couldn't find anything normative in the spec on a quick glimpse. Going to ask on Naga Matrix chat.

**Comment by emilk** [comment]
I get this warning in Firefox:

```
WebGL warning: getSyncParameter: ClientWaitSync must return TIMEOUT_EXPIRED until control has returned to the user agent's main loop. (only warns once)
```

No such warning on chromium.

**Comment by emilk** [comment]
This works great, good job! ⭐

**Review by emilk — APPROVED** [review]

**Comment by Wumpf** [comment]
> I get this warning in Firefox:
> 
> ```
> WebGL warning: getSyncParameter: ClientWaitSync must return TIMEOUT_EXPIRED until control has returned to the user agent's main loop. (only warns once)
> ```
> 
> No such warning on chromium.

Filed an issue https://github.com/gfx-rs/wgpu/issues/3064

**Pull request merged**